prometheus/registry.go

```diff
@@ -722,5 +723,23 @@ func (s metricSorter) Less(i, j int) bool {
 			return vi < vj
 		}
 	}
-	return true
+	if s[i].GetTimestampMs() != s[j].GetTimestampMs() {
+		// Ensure metrics are ordered by timestamp.  Treat
+		// a zero timestamp as "now" in case we have
+		// untimestamped metrics mixed with timestamped
+		// metrics.
+		ti, tj := s[i].GetTimestampMs(), s[j].GetTimestampMs()
+		if ti == 0 {
```

Comment:
Two things here:

(1) problem of an unset timestamp

It's a general issue how to distinguish an unset timestamp and a timestamp == 0, especially as the latter might appear as the former after a serialization round-trip. We are considering using MinInt64 as default value in the protobuf definition.

To not break in that case, you should check for `s[i].TimestampMs == nil` to determine if the timestamp is unset.

(2) problem of "now"

With calling `time.Now()`, the sorting depends on the time of the call (and it might even change during sorting). That's dangerous. I'd suggest to sort an unset timestamp always at the beginning or always at the end.